src/imap.rs

```diff
@@ -361,7 +363,20 @@ impl Imap {
                     Client::connect_secure(context, imap_server, imap_port, self.strict_tls).await
                 }
             };
+            const TIMEOUT_MS: u64 = 120_000;
+            const BACKOFF_MIN_MS: u64 = 1000;
+            let elapsed_ms: u64 = time_elapsed(&conn_start_time).as_millis().try_into()?;
+            if res.is_ok() || elapsed_ms >= TIMEOUT_MS {
+                break res;
```

Comment:
There's a problem though. If `connect()` is called immediately after timing out, the next connection retry is done w/o a sleep. Maybe we could use another `Ratelimit` instance for connection tries, say, `Ratelimit::new(Duration::new(60, 0), 1.0)`, but sleep for `ratelimit.until_can_send() / 60`? I didn't check how this works actually, but if we could achieve a progressive backoff this way, it would be better than now

Reply:
Added `conn_last_try` and `conn_backoff_ms` to `Imap` to solve this. Anyway we already have some state related to retries there (`ratelimit`), so this doesn't change much. And looks like this is simplier than looping inside `Imap::connect()` and doesn't break any tests. If `connect()` starts to sleep for long, any code calling `connect()` several times in a row (this can happen implicitly) breaks.